reorder forward declarations to get rid of C++14-only auto return types

gentryx opened the pull request:
This fixes C++11-only builds (basically anything that's using CUDA). Closes #2718

hkaiser commented:
@gentryx This is roughly what we've had before. Unfortunately this fails compiling on some platforms (sorry, don't remember which) because of incomplete types. Could you please change the PR such that both versions of the code are there, using a pp constant to switch between the two (`HPX_HAVE_CXX11`? - I know this does not exist yet, it would need to be added here: https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_DetectCppDialect.cmake)?

gentryx commented:
@hkaiser Are you referring by "this" to the use of decltype instead of auto for the return types in the customization points? Yeah, I can change that to be configurable, no problem. :-)

hkaiser commented:
@gentry: I meant the whole forward declaration. Also, the use of `auto` without an explicit return type is intentional. So I'd appreciate it if you left the existing code unchanged as an alternative to the code you would like to submit.

gentryx commented:
@hkaiser I think I don't quite understand. What I did was basically two things:

1. I changed the C++14-style use of "auto" as a return type to be C++-11-compliant. I can change that to use the C++14-style if available.

2. I rearranged the order of a ton of class/function definitions to make the C++11-style work (because that's referring to functions and classes that were previously defined after the customization points. This should not affect any platform, right? If it's working for my stricter builds, it should work on other platforms just as well. Or am I missing something?

gentryx commented:
Oh, and I forgot:

3. I removed the forward declaration for the customization points. Do you want me to re-add that to execution_fwd.hpp? I suppose I can do that as well, no problem.

hkaiser commented:
@gentry: 

- In C++11 you can use just `auto` without explicit return type specification if the body of the function is just a single return statement.
- The reordering you propose is exactly what we started off with but we had to change it to use the forward declaration as other compilers were complaining about incomplete types in C++14 mode
- I'd like to ask you to (optionally) leave the code as it is right now and introduce your fix via a preprocessor constant which switches between the two implementations

Thanks!

gentryx commented:
@hkaiser: I've added preprocessor guards so that the customization points will again use forward declarations and "auto" if C++14 is available, but will fall back to a C++11-compatible style if necessary.

I've successfully tested this build with GCC 4.9.4 with and without CUDA 8.0 for the C++11-mode and with GCC 6.3.0 for a C++14 build. Please let me know if you want me to try further compiler versions. :-)

hkaiser commented on cmake/tests/cxx14_auto.cpp:
```diff
@@ -0,0 +1,20 @@
+{
+    T j = 1 + j;
+    return j;
+}
```
This is not C++11. For this, the body of the function has to consist of only the return statement.

hkaiser replied:
Let me rephrase this. If the feature test is for a C++14 feature (which is fine), why do you use the result of this feature test to add a workaround for a missing C++11 capability of a particular compiler? Shouldn't we rather feature test for the particular missing capabilities und use this instead to decide whether to activate the workaround?

gentryx replied:
Ah, got it. You're right, this is apparently a bug in NVCC. However, CMake isn't using NVCC to test for compiler flags (but rather the default C++ compiler), so a test here for that specific defect likely wouldn't yield us much.

If you object to this test as it is in the PR then I think it would be fine to just tell CUDA users to manually set "-DWITH_CXX11_AUTO=false". In that case I'd remove the new C++14 auto test. Your thoughts?

hkaiser replied:
I'm not opposed to this feature test. I'm opposed to using the results of this test to detect a NVCC problem related to a C++11 feature. Couldn't we just force HPX_WITH_CXX11_AUTO=OFF whenever HPX is configured with CUDA/NVCC enabled?

gentryx replied:
At CMake time we don't know if the user is going to use nvcc for their code.

I can drop the additional (misleading) CXX14 CMake test and then users can manually disable CXX11 auto support. Would that be better?

hkaiser commented:
@gentry: should those changes be applied to the execution information and time execution customization points as well: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/execution_information_fwd.hpp, https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/timed_execution_fwd.hpp)?

hkaiser commented:
> should those changes be applied to the execution information and timed execution customization points as well: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/execution_information_fwd.hpp and https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/timed_execution_fwd.hpp?

@gentryx: to clarify: you have applied different methods of resolving the issue at hand for the base execution customization points compared to how you resolved it for the timed execution and informational customization points. Why?

Also, the original code was using the `auto` return type deduction for functions on purpose (without explicit trailing return type). This improves error messages as the `customization_point<>::operator()()` implementations are not SFINAE'd out in case of some underlying problems. You propose to unconditionally replace this with explicit return type deductions (`-> decltype(...)`) for the customization points related to the timed execution and informational facilities. I understand that we have to add explicit return type deduction for some compilers, but we should leave the original code in place for conforming ones.

You have also removed the headers `timed_execution_fwd.hpp` and execution_information_fwd.hpp` from the documentation without re-adding the corresponding headers (which now have the doxygen comments). This removes those customization points from the generated docs. A similar problem now occurs related to the doxygen comments for the main customization points. Those comments are now in a new header file which is not visible to the documentation tool chain, again removing them from the generated docs.

gentryx commented:
@hkaiser I added explicit return types (via decltype(...)) for customization_point because of NVCC's shortcomings WRT auto return types. This seems to be the only solution that works with NVCC (so apparently NVCC is not standard compliant). However, this is conditional on the CMake check for C++14 auto support, and if such support is found the code after the preprocessor is identical to the original code. As mentioned above I can change the conditional from "C++14 auto" to "C++11 auto".

If this approach is fine for you then I can follow the same pattern for timed_execution_fwd.hpp and execution_information_fwd.hpp. I had these headers removed as my builds (g++ 4.9.4 with CUDA 8.0 g++ 6.2.0 and clang 4.0) work just fine with moving those functions into the corresponding headers, but I hold no stake in this, I'm fine with doing the same as I did for execution.hpp/execution_fwd.hpp.

gentryx commented:
OK, looks like CircleCI is fine with the PR (the error it's throwing seems related to the Perl install?). :-)

Please advise on how to procedd WRT timed_execution_fwd.hpp/execution_information_fwd.hpp. Same pattern as for execution_fwd.hpp?

hkaiser commented:
> OK, looks like CircleCI is fine with the PR (the error it's throwing seems related to the Perl install?). :-)

No idea what's wrong there... I have never seen this before.

> Please advise on how to procedd WRT timed_execution_fwd.hpp/execution_information_fwd.hpp. Same pattern as for execution_fwd.hpp?

Yes, please.

hkaiser commented:
@gentryx: are you still working on this?

gentryx commented:
@hkaiser Please take another look (if CircleCI succeeds). I updated this PR to HPX' trunk and implemented it the way we discussed earlier.

Tested with g++ 4.9 and CUDA, g++ 6.3 and clang++ 4.0.

gentryx commented:
@hkaiser Looks like this one went through successfully. Please take another look.

hkaiser commented on cmake/HPX_AddConfigTest.cmake:
```diff
@@ -208,7 +209,7 @@ macro(hpx_check_for_cxx11_auto)
     SOURCE cmake/tests/cxx11_auto.cpp
     FILE ${ARGN}
-    CMAKECXXFEATURE cxx_auto_type)
+    CMAKECXXFEATURE cxx11_auto_type)
```
There is no C++ feature detection property in cmake called `cxx11_auto_type` (please see https://cmake.org/cmake/help/v3.9/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

hkaiser commented on cmake/HPX_AddConfigTest.cmake:
```diff
@@ -511,6 +512,14 @@ macro(hpx_check_for_cxx11_std_unordered_set)
+  add_hpx_config_test(HPX_WITH_CXX14_AUTO
+    SOURCE cmake/tests/cxx14_auto.cpp
+    FILE ${ARGN}
+    CMAKECXXFEATURE cxx14_auto_type)
```
There is no C++ feature detection property in cmake called `cxx14_auto_type` (please see https://cmake.org/cmake/help/v3.9/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

hkaiser commented on hpx/exception_info.hpp:
```diff
@@ -73,10 +74,10 @@ namespace hpx
         {
         public:
-            template <typename ...Tags, typename ...Types>
+            template <typename ...ErrorInfo>
```
What is the rationale of changing this file?

gentryx replied:
NVCC couldn'T compile this.

hkaiser commented on hpx/parallel/executors/execution_fwd.hpp:
```diff
@@ -67,6 +68,7 @@ namespace hpx { namespace parallel { namespace execution
             }
         };
 
+#ifdef HPX_HAVE_CXX14_AUTO
```
Please don't use a macro named `_CXX14_...` to guard code which is C++11 compliant.

gentryx replied:
ACK, will fix.

gentryx commented:
I've changed the PR according to your comments, build failed due to the usual Perl install error. PTAL :-)

gentryx commented:
@hkaiser Is there anything else you want changed for this PR or is this good to go?

hkaiser commented:
@gentry: frankly I'm not too happy to relax the requirement for HPX to depend on `auto` in general. This is a problem specific to one compiler after all. But at this point I'm not sure how to proceed. Please give me a day or two to have a closer look and to play with this. 

The main problem is that we don't run the feature tests for the compiler used for compiling CUDA but for the base compiler NVCC is using to generate code...

gentryx commented:
@hkaiser In this PR I'm using the CMake auto test only to switch between different definitions for the affected classes. Would you prefer if I introduced a second CMake test so that the current one is mandatory again and the second one can be overridden?

hkaiser commented:
> Would you prefer if I introduced a second CMake test so that the current one is mandatory again and the second one can be overridden?

@gentryx , yah, that might be a good compromise.

Naios commented on hpx/parallel/executors/execution_information.hpp:
```diff
@@ -270,15 +323,99 @@ namespace hpx { namespace parallel { namespace execution
+        /// \note This calls exec.set_scheduler_mode(mode) if it exists;
+        ///       otherwise it does nothing.
+        ///
+        constexpr detail::customization_point<
```
`HPX_CONSTEXPR`

gentryx replied:
Yeah, maybe better do that in a separate PR instead of getting sidetracked here :-)

gentryx commented:
@hkaiser I've reverted the change to cmake/HPX_PerformCxxFeatureTests.cmake and introduced a new CMake option HPX_WITH_CXX11_AUTO_RETURN_VALUE which is only used in the code I've changed. The option is on by default, but can be overridden by the user.

gentryx commented:
Builds are successful (modulo the usual Perl woes).

hkaiser commented:
@gentryx: thanks a lot, I'll have a look later today. I still don't understand where those Perl problems come from, though - this happens on this branch only.

gentryx commented:
@hkaiser nag nag

hkaiser commented:
@gentryx: sorry for the delay - I'll do it asap.

gentryx commented:
@hkaiser I can understand that you're conservative about merging a failing build since you don't want to break the trunk. But I'd like to add that my HPX+LGD builds have been broken now for quite a while, or in other words: the HPX trunk is broken for CUDA since then.

My best guess is that for my build I see "CIRCLE_BUILD_IMAGE=ubuntu-14.04" while in https://circleci.com/gh/STEllAR-GROUP/hpx/7572?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link I see "CIRCLE_BUILD_IMAGE=ubuntu-12.04", so I'd say the build failures are unrelated to my PR and simply a sign that HPX builds with newer CircleCI images will fail.

gentryx commented:
I've now switched my HPX builds to use Ubuntu 12.04 if that's required to get this merged. You should file a separate ticket to make sure HPX builds with 14.04 too because I assume that 12.04 will some day no longer be available.

hkaiser commented:
@gentryx: sorry for going back and forth on this, but wouldn't it be sufficient to apply this patch to the main CMakeLists.txt in order to disable the functionality for NVCC only:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fd416de2be..fcfefe5472 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1084,6 +1084,7 @@ if(HPX_WITH_CUDA)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};--default-stream per-thread)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-lcudadevrt)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-rdc=true)
+    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-DHPX_HAVE_CXX11_AUTO_RETURN_VALUE)
   else()
     hpx_add_target_compile_option(-DBOOST_THREAD_USES_MOVE)
     hpx_add_target_compile_option(--cuda-path=${CUDA_TOOLKIT_ROOT_DIR})
@@ -1112,13 +1113,6 @@ else()
   endif()
 endif()
 
-hpx_option(HPX_WITH_CXX11_AUTO_RETURN_VALUE BOOL
-  "Enable the use of auto as a return value in some places. Overriding this flag is only necessary if the C++ compiler is not standard compliant, e.g. nvcc."
-  ON ADVANCED)
-if(HPX_WITH_CXX11_AUTO_RETURN_VALUE)
-  hpx_add_config_define(HPX_HAVE_CXX11_AUTO_RETURN_VALUE)
-endif()
-
 ################################################################################
 # check for miscellaneous things
 ################################################################################

```

hkaiser commented:
@gentryx: also, may I suggest to rewrite the patch in the following fashion:
```
diff --git a/hpx/parallel/executors/execution_fwd.hpp b/hpx/parallel/executors/execution_fwd.hpp
index 5df2085aa9..1fc00821e0 100644
--- a/hpx/parallel/executors/execution_fwd.hpp
+++ b/hpx/parallel/executors/execution_fwd.hpp
@@ -307,29 +307,29 @@ namespace hpx { namespace parallel { namespace execution
 
         ///////////////////////////////////////////////////////////////////////
         // async_execute customization point
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+#ifdef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+        template <typename Executor, typename F, typename ... Ts>
+        HPX_FORCEINLINE
+        auto customization_point<async_execute_tag>::operator()(
+            Executor && exec, F && f, Ts &&... ts) const
+        {
+            return async_execute(std::forward<Executor>(exec),
+                std::forward<F>(f), std::forward<Ts>(ts)...);
+        }
+#else
         template <>
         struct customization_point<async_execute_tag>
         {
         public:
-#endif
             template <typename Executor, typename F, typename ... Ts>
             HPX_FORCEINLINE
-            auto
-#ifdef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
-            customization_point<async_execute_tag>::
-#endif
-            operator()(
-                Executor && exec, F && f, Ts &&... ts) const
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+            auto operator()(Executor && exec, F && f, Ts &&... ts) const
             -> decltype(async_execute(std::forward<Executor>(exec),
                     std::forward<F>(f), std::forward<Ts>(ts)...))
-#endif
             {
                 return async_execute(std::forward<Executor>(exec),
                     std::forward<F>(f), std::forward<Ts>(ts)...);
             }
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
         };
 #endif
 
```

yes, this adds some duplication of code but significantly enhances readability.

hkaiser commented:
Otherwise, this LGTM, thanks a lot for your persistence and patience!

hkaiser commented:
>  I've now switched my HPX builds to use Ubuntu 12.04 if that's required to get this merged. You should file a separate ticket to make sure HPX builds with 14.04 too because I assume that 12.04 will some day no longer be available.

I don't think this is an OS version issue. It looks like the newer OS version does not come with the Perl module which is causing the problems. In the end it does not really matter as the docker image is maintained by us, so nobody will change the used OS without us knowing.

gentryx commented:
@hkaiser 

- WRT setting CUDA_NVCC_FLAGS: yes, that would be one solution. It's got side effects though as you then assume that the user enables HPX' components for CUDA (-DHPX_WITH_CUDA=true), which may not be what a user wants. Example: If I set this option then HPX will export the flag "--default-stream per-thread" via CUDA_NVCC_FLAGS. This can break existing CUDA code in subtle ways. I'd say: this is opening another can of worms.

- WRT code duplication vs. readability: TBH since CUDA builds don't seem to be in your build rotation I specifically wanted to avoid having a header where someone could change one code path and leave the other unchanged.

- WRT OS versions: yes, it's more precise to say that it's not Ubuntu 14.04 that's causing the breakage here, but the image of Ubuntu 14.04 that Circle CI is using. The image is selected in the settings of each user, the docker image is more like an overlay, right? At least we didn't quite spot this setting in my builds weeks.

hkaiser commented:
> WRT setting CUDA_NVCC_FLAGS: yes, that would be one solution. It's got side effects though as you then assume that the user enables HPX' components for CUDA (-DHPX_WITH_CUDA=true), which may not be what a user wants. Example: If I set this option then HPX will export the flag "--default-stream per-thread" via CUDA_NVCC_FLAGS. This can break existing CUDA code in subtle ways. I'd say: this is opening another can of worms.

Yes, you're right. Could we at least make this configuration setting `ON` whenever CUDA/NVCC is enabled, leaving it ``OFF` by default? Otherwise the user would have to specify two configuration options in order to build HPX with CUDA.

> WRT code duplication vs. readability: TBH since CUDA builds don't seem to be in your build rotation I specifically wanted to avoid having a header where someone could change one code path and leave the other unchanged.

FWIW, we have CUDA builds, but those are broken for weeks because of this issue .... ;-)

Adding a comment warning the user to change both code path when needed should take care of this. The way it is right now the code is close to impossible to grasp, at least for me.

gentryx commented:
K, I've refactored the patch for readability as you suggested. The new CXX11 auto option is now off by default if HPX_WITH_CUDA is selected.

hkaiser commented:
> K, I've refactored the patch for readability as you suggested. 

Thanks, perfect!

> The new CXX11 auto option is now off by default if HPX_WITH_CUDA is selected.

Uhh, shouldn't it be `ON` by default if CUDA is selected, but `OFF` by default otherwise?

hkaiser commented:
@gentryx: uhh, I moved this over to the branch https://github.com/STEllAR-GROUP/hpx/tree/gentryx-master in the main repo and fixed some things. I'll try to reintegrate your work with that branch tomorrow. Thanks!

gentryx commented:
@hkaiser Okidoki. I've just pushed some changes to please inspect. And yes: the option should be on by default and off if CUDA is being used as NVCC doesn't support it.

gentryx commented:
@hkaiser Regarding your branch gentryx-master: the default value for the option needs to be inversed: on by default, off if CUDA (with gcc) is selected.

hkaiser commented:
@gentry: ok, I'll verify again.

hkaiser commented:
@gentryx I changed the default for the option (thanks for pointing this out) and reintegrated all of your recent changes into https://github.com/STEllAR-GROUP/hpx/tree/gentryx-master (please check, I hope I have not missed anything). Once circleci passes I'm going to merge things from there.

gentryx commented:
@hkaiser Looks good so far. I've merged the gentryx-master branch in this PR and added another commit to mask some tests that won't compile in CXX11-mode (as they expect auto-typed lambda parameters).

hkaiser commented:
@gentryx: thanks so much!

gentryx commented:
I have no clue what the error message means that CircleCI is showing. Does this block the merge?

hkaiser commented:
@gentry: I think this is happening because the build - for some reason - runs under your account, not under the STEllAR-GROUP account. I think we're fine merging anyways.

gentryx commented:
Woohoo! :-)
